src/freertos_drivers/common/BitBangI2C.cxx

```diff
@@ -0,0 +1,353 @@
+                    if (stop_)
+                    {
+                        state_ = State::STOP;
+                        stateStop_ = StateStop::FIRST;
+                    }
+                    else
+                    {
+                        exit = true;
+                    }
+                }
+            }
+            break;
+        case State::STOP:
+            exit = state_stop();
+            break;
+    }
+
+    if (exit)
+    {
+        disableTick_();
+        int woken = 0;
+        sem_.post_from_isr(&woken);
+        os_isr_exit_yield_test(woken);
+    }
+}
+
+//
+// BitBangI2C::state_start()
+//
+__attribute__((optimize("-O3")))
+bool BitBangI2C::state_start()
+{
+    switch (stateStart_)
+    {
+        // start sequence
+        case StateStart::SDA_SET:
+            gpio_set(sda_);
+            ++stateStart_;
+            break;
+        case StateStart::SCL_SET:
+            gpio_set(scl_);
+            ++stateStart_;
+            break;
+        case StateStart::SDA_CLR:
+            gpio_clr(sda_);
+            ++stateStart_;
+            break;
+        case StateStart::SCL_CLR:
+            gpio_clr(scl_);
+            return true;
+    }
+    return false;
+}
+
+//
+// BitBangI2C::state_stop()
+//
+__attribute__((optimize("-O3")))
+bool BitBangI2C::state_stop()
+{
+    switch (stateStop_)
+    {
+        case StateStop::SDA_CLR:
+            gpio_clr(sda_);
+            ++stateStop_;
+            break;
+        case StateStop::SCL_SET:
+            gpio_set(scl_);
+            ++stateStop_;
+            break;
+        case StateStop::SDA_SET:
+            gpio_set(sda_);
+            stop_ = false;
+            return true; // exit
+    }
+    return false;
+}
+
+//
+// BitBangI2C::state_tx()
+//
+__attribute__((optimize("-O3")))
+bool BitBangI2C::state_tx(uint8_t data)
+{
+    switch(stateTx_)
+    {
+        case StateTx::DATA_7_SCL_SET:
+        case StateTx::DATA_6_SCL_SET:
+        case StateTx::DATA_5_SCL_SET:
+        case StateTx::DATA_4_SCL_SET:
+        case StateTx::DATA_3_SCL_SET:
+        case StateTx::DATA_2_SCL_SET:
+        case StateTx::DATA_1_SCL_SET:
+        case StateTx::DATA_0_SCL_SET:
+        {
+            uint8_t mask = 0x80 >>
+                ((static_cast<int>(stateTx_) -
+                  static_cast<int>(StateTx::DATA_7_SCL_SET)) >> 1);
+            if (data & mask)
+            {
+                gpio_set(sda_);
+            }
+            else
+            {
+                gpio_clr(sda_);
+            }
+            gpio_set(scl_);
+            ++stateTx_;
+            break;
+        }
+        case StateTx::DATA_7_SCL_CLR:
+        case StateTx::DATA_6_SCL_CLR:
+        case StateTx::DATA_5_SCL_CLR:
+        case StateTx::DATA_4_SCL_CLR:
+        case StateTx::DATA_3_SCL_CLR:
+        case StateTx::DATA_2_SCL_CLR:
+        case StateTx::DATA_1_SCL_CLR:
+        case StateTx::DATA_0_SCL_CLR:
+            gpio_clr(scl_);
```

Comment:
```
// I2C is specified such that the data on the SDA line must be stable during the high period of the clock, 
// and the data line can only change when SCL is Low.
//
// This means that the sequence always has to be 
// 1a. SCL := low
// 1b. set/clear SDA
// 2. wait a cycle for lines to settle
// 3a. SCL := high
// 3b. this edge is when the receiver samples
// 4. wait a cycle, lines are stable
 ```

This also means that we need to start the state machine by clearing the clock.

Reply:
![image](https://user-images.githubusercontent.com/3719690/228525543-b2b9bf12-2fd4-4afc-a032-319dc146c024.png)